File: week04/127-WordLadder.java
class Solution {
    public int ladderLength(String start, String end, List<String> wordList) {
        Set<String> dic = new HashSet<>();
        for (String word : wordList) {
            dic.add(word);
        }
        
        int length = 1;
        Queue<String> queue = new LinkedList<>();
        Set<String> hash = new HashSet<>();
        queue.offer(start);
        hash.add(start);
        
        while (!queue.isEmpty()) {
            length++;
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String word = queue.poll();
                
                for (String nextWord : getNextWords(word, dic)) {
                    if (hash.contains(nextWord)) continue;
                    if (nextWord.equals(end)) return length;
                    
                    queue.offer(nextWord);
                    hash.add(nextWord);
                }
            }
        }
        return 0;
    }
    
    private String replace(String s, int index, char c) {
        char[] chars = s.toCharArray();
        chars[index] = c;
        return new String(chars);
    }
    
    private ArrayList<String> getNextWords(String word, Set<String> dic) {
        ArrayList<String> nextWords = new ArrayList<>();
        for (char c = 'a'; c <= 'z'; c++) {
            for (int i = 0; i < word.length(); i++) {
                if (word.charAt(i) == c) continue;
                String nextWord = replace(word, i, c);
                
                if (dic.contains(nextWord)) nextWords.add(nextWord);
            }
        }
        return nextWords;
    }
}

File: week04/127-WordLadder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(start, end, wordList) {
    let que = [start], distance = new Map(), dic = new Set(wordList);
    distance.set(start, 1); // distance 记录开始到当前单词的距离
    
    while (que.length) {
        let word = que.shift();
        if(word === end) return distance.get(word); 
        
        for (let nextWord of getNextWords(word)) {
            if (!dic.has(nextWord) || distance.has(nextWord))
                continue;
            que.push(nextWord);
            distance.set(nextWord, distance.get(word) + 1);
        }
    }
    return 0;
};

const getNextWords = (word) => {
    let words = [];
    for (let i = 0; i < word.length; i++) {
        let left = word.slice(0,i) , right = word.slice(i + 1);
        
        for(let c of 'abcdefghijklmnopqrstuvwxyz') {
            if (word[i] === c) continue;
            words.push(left + c + right);
        }
    }
    return words;
}
